design/src/rfcs/rfc0025_better_constraint_violations.md

```diff
@@ -0,0 +1,511 @@
+        // Check constraints and collect violations.
+        ...
+    }
+}
+```
+
+- The main reason for wrapping a vector in `ConstraintViolations` as opposed to
+  directly returning the vector is forwards-compatibility: we may want to
+  expand `ConstraintViolations` with conveniences.
+- If the constrained type can only ever yield a single violation, we will
+  dispense with `ConstraintViolations` and keep directly returning the
+  `crate::model::shape_name::ConstraintViolation` type.
+
+We will analogously introduce a `ConstraintViolationExceptions` type that
+represents a collection of constraint violations that can occur _within the
+framework's request deserialization code_. This type will be `pub(crate)` and
+will be the one the framework will map to Smithy's `ValidationException` that
+eventually gets serialized into the response.
+
+### Unresolved questions
+
+#### Collecting constraint violations may constitute a DOS attack vector
+
+This is a problem that _already_ exists as of writing, but that collecting
+constraint violations highlights, so it is a good opportunity, from a
+pedagogical perspective, to explain it here. Consider the following model:
+
+```smithy
+@length(max: 3)
+list ListOfPatternStrings {
+    member: PatternString
+}
+
+@pattern("expensive regex to evaluate")
+string PatternString
+```
+
+Our implementation currently enforces constraints _from the leaf to the root_:
+when enforcing the `@length` constraint, the `TryFrom` constructor the server
+framework uses gets a `Vec<String>` and _first_ checks the members adhere to
+the `@pattern` trait, and only _after_ is the `@length` trait checked. This
+means that if a client sends a request with `n >>> 3` list members, the
+expensive check runs `n` times, when a constant-time check inspecting the
+length of the input vector would have sufficed to reject the request.
+Additionally, we may want to avoid serializing `n` `ValidationExceptionField`s
+due to performance concerns.
+
+- A possibility to circumvent this is making the `@length` validator special,
+  having it bound the other validators via effectively permuting the order of
+  the checks and thus short-circuiting.
+  * In general, it's unclear what constraint traits should cause
+    short-circuiting. A probably reasonable rule of thumb is to include
+    traits that can be attached directly to aggregate shapes: as of writing,
+    that would be `@uniqueItems` on list shapes and `@length` on list shapes.
+- Another possiblity is to _do nothing_ and value _complete_ validation
+  exception response messages over trying to mitigate this with special
+  handling. One could argue that these kind of DOS attack vectors should be
+  taken care of with a separate solution e.g. a layer that bounds a request
+  body's size to a reasonable default (see [how Axum added
+  this](https://github.com/tokio-rs/axum/pull/1420)). We will provide a similar
+  request body limiting mechanism regardless.
```

Comment:
I think the solution Axum is using is too simplistic for our usecase. We may have customers that may want a bigger body size or even a completely unconstrained one. 

We could find a decent default and make it configurable for advanced users..

Reply:
The RFC now advocates that we should add both solutions after the rewrite: https://github.com/awslabs/smithy-rs/pull/2040/commits/5033c7af05e7e1aa9fe2308cf52dd27fe0c3deee.